### PR TITLE
Feature/integrate 1d boost

### PR DESCRIPTION
**Pull request by bbbales2**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

This pull req. switches to the Boost 1D integrator and addresses a bunch of the interface issues brought up here: https://github.com/stan-dev/math/pull/759

Notes:
1. Boost integrators are here

2. Boost integrators use a different definition of tolerance than the previous integrator

3. Limits at infinity are technically included, but it's very easy for autodiff to fail. The integrators will evaluate the functions near or at the numeric_limits, so numbers quickly become infs and math with infs turns into NaNs. Integrating a normal pdf from negative to positive infinity with a standard deviation that's a parameter won't work for instance (if the standard deviation is fixed it's fine -- the problem is in the integral of the automatically differentiated function)

Based on this I'm not sure if we should do limits at infinity. The autodiff is definitely going to cause bunches of issues here. Thoughts?

4. I did not end up doing the two argument function. Two things:
  a. Boost recommends splitting integrals that cross zero into two. That's easy to automate, and so I did it. That means that xc can be a distance two up to three different points now, so it's a little weirder to work with.
  b. I think it's easy enough to replicate the behavior by either hard coding things in or passing them in as data (x_r)

Anything I'm missing here?

5. The limits of integration cannot be vars in this integration.

@bgoodri can you review this? The easiest place to start will be stan/math/rev/arr/functor/integrate_1d.hpp

#### Intended Effect:

Switch to Boost for 1d integration and allow infinite integrals.

#### How to Verify:

./runTests.py test/unit/math/prim/arr/functor/integrate_1d_test.cpp
./runTests.py test/unit/math/rev/arr/functor/integrate_1d_test.cpp

#### Side Effects:

Changed interpretation of tolerance for 1d integration. Removed ability to pass in custom gradient functions

#### Documentation:

In the doxygen!

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bgoodri** [comment]
We should be switching to sinh_tanh or exp_tanh when the limits are infinite.

I don't understand how you are proposing to handle a beta integral for example without the two argument version.

It seems like allowing the limits of integration to be (finite) vars should be easy enough if we do like value_of(b) and then handle the derivative analytically.

**Comment by bbbales2** [comment]
I used tanh_sinh for [a, b], sinh_sinh with one side at infinity, and exp_sinh for [-inf, inf]. I followed the Boost docs pretty closely.

Maybe I misunderstood where the problem with the xc. In the Boost example, I thought it came down to 1 - x^2 will be hard because subtraction from the number 1 is limited by the number of digits you have in double. So if x is about 1e-9, then 1 - x^2 is probly just one, but evaluating it like (1 - x) * (1 + x) won't be as difficult because we can take x down to probly 1e-17 before (1 - x) == 1 and (1 + x) == 1. I don't see the difficulty in the beta function here: https://en.wikipedia.org/wiki/Beta_function as long as we're talking about integer x and y? What am I missing?

Well if the endpoint is a var, then we need all the gradients da/dtheta and db/dtheta. We don't have those analytically. Technically we could get them by clearing the autodiff stack and evaluating a reverse mode .grad(), but it could be pretty inefficient if there's unrelated stuff sitting on the stack.

**Comment by bbbales2** [comment]
Huh, yeah, I checked. Integrating the beta numerically seems super hard for even x = 0.1, y = 0.1. The implementation here fails. But I still don't see how the xc sorta thing is going to enable a parameterization we can't accomplish by either knowing the problem or passing extra info through x_r.

**Comment by bgoodri** [comment]
What info are you proposing passing through x_r to solve the U-shaped beta
integral problem?

**Comment by bbbales2** [comment]
Well, what info do you need? Cause I'm saying I can compute the xc Boost will give you on the fly by just saying:

```
xc = (x > 0.5) ? 1 - x : x;
```

And there I don't even need x_r cause the domain is a simple [0, 1].

**Comment by bgoodri** [comment]
But computing the 1 - x is where the precision gets lost for x near 1.

**Comment by bbbales2** [comment]
Hmm, you're right. Checked with your reference code. Just 1 - x is a problem too. Okeedokee that's easy enough fix. I'll add beta function integrals to the tests.

Should I just remove the code that automatically splits the integral at 0? Boost folks say to do it, but it'd make interpreting xc slightly trickier.

**Comment by bgoodri** [comment]
I would say auto split at 0 but doc the fact that their two-argument
function will be called differently in the two parts.

**Comment by bbbales2** [comment]
Sounds like a plan

**Comment by bgoodri** [comment]
Ready @bbbales2 ?

**Comment by bbbales2** [comment]
Yup, give it a go. The beta gradients seemed much harder to get right than
the functions themselves.

**Comment by bgoodri** [comment]
Okay. I imagine it is possible for the integrand to be an analytic function
but not the gradient of the integrand with respect to the parameters, in
which case the convergence of the former would be much faster than the
convergence of the latter.

**Comment by bgoodri** [comment]
Do we want to make E_2(h, 1) from equation (14) in https://www.davidhbailey.com/dhbpapers/em-error.pdf available as well in order to compute the error more rigorously?

**Comment by bbbales2** [comment]
Hmm, does that require the 2nd derivative of f though? I suppose we could finite difference but seems like the numbers would get super crude.

We could certainly pipe more error information around at the stan-math level. At the actual Stan level though, I dunno. If we had tuples, maybe. How'd you see this extra information getting handled?

**Comment by bgoodri** [comment]
I was thinking a separate function that called `stan::math::gradient` and `stan::math::hessian` on the user-defined function to evaluate their E_2(h, 1) and returned the error estimate in case someone wanted to check their accuracy more.

**Comment by bob-carpenter** [comment]
We're not through to trusting the Hessians through our matrix lib.  We need to fix that ASAP so we can do this kind of thing.  Hessians are O(N) more expensive than log densities to compute for N parameters plus a big constant multiplier for all the forward-mode stuff.

**Comment by bgoodri** [comment]
Do we trust it for univariate functions?

**Comment by bgoodri** [comment]
@bbbales2 What were you saying about autodiffing to infinity being a problem? I added this to test/unit/math/rev/arr/functor/integrate_1d_test.cpp locally and it is fine.
```
+struct f12 {
+  template <typename T1, typename T2>
+  inline typename stan::return_type<T1, T2>::type operator()(
+      const T1& x, const T1& xc, const std::vector<T2>& theta,
+      const std::vector<double>& x_r, const std::vector<int>& x_i,
+      std::ostream& msgs) const {
+    T2 sigma = theta[0];
+    return exp(-0.5 * stan::math::square(x / sigma))
+           / (sigma * sqrt(2.0 * stan::math::pi()));
+  }
+};
```
The Boost doc does say you have to do your own limits inside the function sometimes, like instead of doing `x * log(x)` you have to do `x > 0 ? x * log(x) : 0`.

**Comment by bob-carpenter** [comment]
Yes, all the univariate functions and C++ built ins have been tested thoroughly.   

There remains some danger they'll participate in multivariate functions, like the sum of a scalar-matrix multiply.

At some point, we're going to have to "turn on" forward-mode autodiff for general use.  Since the forward mode stuff only uses the templated implementations, everything should be fine if it builds.

**Comment by bbbales2** [comment]
Hmm. I tried a Gaussian

Maybe I wrote it badly. I definitely started with x^2 / sigma^2 though so that’s different.

**Comment by bbbales2** [comment]
Good find though I’d prefer that things work :/

**Comment by bbbales2** [comment]
Huh, good point. The limits should be fine as vars. I’ll try to fix that tomorrow AM

**Comment by bbbales2** [comment]
Cool idea. An elegant way to set it up eludes me now though. Want this as a separate issue or want to do it now?

**Comment by bgoodri** [comment]
I wasn't thinking elegant, although it can be done with lambda functions. I
wouldn't say that it is necessary for this PR, but it would possibly catch
some edge cases.

**Comment by bbbales2** [comment]
I tried a lambda solution, but what function gets handed to integrate_1d needs to be templatable. The regular integration does math with doubles, and then the gradients use vars.

I guess it's just 20 or so functors that'd need defined, and probly most of the thought would go into what parameters would be hard to integrate. Make a separate issue for this if you want to see this. It would be interesting. Almost definitely some integral is going to be hard and break though, so I don't wanna stall this pull req. while I debug the numerics of whatever weird lpdf.

**Comment by bgoodri** [comment]
I thought that was legal in C++14

https://solarianprogrammer.com/2014/08/28/cpp-14-lambda-tutorial/

**Comment by sakrejda** [comment]
I think you might be getting distracted.. the PR is big enough as is...

**Comment by bob-carpenter** [comment]
I think this is just an issue of how to generate some simple tests---it's not adding new features.  I agree with @bbbales2 that we don't want to hold up the PR for edge cases, but it would be nice to test expected failure behavior when the algorithm can't get the right answer.  How much is required will be up to the code reviewer.

I tried the example that @bgoodri linked, and the generalized lambda works for me:

```
#include<iostream>
#include<complex>

int main() {
    auto func = [](auto input) { return -0.5 * linput * input; };
    std::cout << func(10) << std::endl;
    std::cout << func(2.345) << std::endl;
    std::cout << func(std::complex<double>(3, -2)) << std::endl;
    return 0;
}
```

and it works with the compiler settings baked into our makefile scripts:

```
~/temp2$ clang++ -std=c++1y template-lambda.cpp 
~/temp2$ ./a.out
-50
-2.74951
(-2.5,6)
~/temp2$ 
```

This is with:

```
~/temp2$ clang++ --version
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

P.S.  It also works with closures:

```
  double mu = -1.2;
  double sigma = 0.75;
  auto f = [&](auto y) { return -0.5 * (y - mu) * (y - mu) / (sigma * sigma); };
```

**Comment by bbbales2** [comment]
Cool beans, what's the latest C++ version we're using on the project these days?

**Comment by bob-carpenter** [comment]
Everything supported by the recommended version of Rtools.  Not exactly sure what that is.

**Comment by bgoodri** [comment]
4.9.3

**Comment by bbbales2** [comment]
@bgoodri, you know which pull req. doesn't require you to wrassle with gcc 4.9.3 :P?

**Comment by bgoodri** [comment]
I think the problem is with

https://github.com/stan-dev/math/pull/913/files#diff-e357c4b0c3d192168d5fbc2487103b31R111

where it does `value_of(theta)`. If you then call one of the functions in `stan/math/prim/scal/prob` that does not calculate the derivatives when its arguments are not vars, then I can see why they would be nan. If so, that is a serious problem, but I am not sure how to fix it.

File: stan/math/rev/arr/functor/integrate_1d.hpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                            const std::vector<int>& x_i, size_t n,
+                            std::ostream& msgs) {
+  double gradient = 0.0;
+  start_nested();
+  std::vector<var> theta_var(theta_vals.size());
+  try {
+    for (size_t i = 0; i < theta_vals.size(); i++)
+      theta_var[i] = theta_vals[i];
+    f(x, xc, theta_var, x_r, x_i, msgs).grad();
+    gradient = theta_var[n].adj();
```
Comment (bgoodri) [inline stan/math/rev/arr/functor/integrate_1d.hpp]: We need to have a `if (is_nan(gradient))` here and throw an informative exception if necessary. The question is can we get away with changing gradient from nan to zero if the value of the function is zero?

**Comment by bbbales2** [comment]
@bgoodri Good catch. Let's go with throwing an error instead of swapping in zeros for the nans though. Agreed?

You can avoid this sort of thing with:
```
if(abs(x) > 1e300) {
  return 0.0;
} else {
  return exp(stan::math::normal_lpdf(x, mu, sigma));
}
```

which is pretty easy to tell people how to do.

This is just Boost evaluating things at the limits of double precision. It's really easy (depending on how things are implemented) to get infs, and then inf / inf or inf - inf will be a NaN and that's it.

**Comment by bgoodri** [comment]
You can also accomplish the same thing by integrating over a huge but finite interval. The question is whether we want to force users to do that and also whether we want to kill the sampler just because of the gradient overflowing.

It is really close. In this particular context, the fact that you are calling `integrate_1d` over an unbounded interval implies that the integrand must approach zero eventually and given the sketchiness of our gradients for extreme values that basically implies this problem is likely to come up. Testing that the function is exactly zero (i.e. underflowed to zero) as opposed to just being near zero seems pretty safe. In a lot of situations, we do not really have a numerically well-defined gradient when the function is underflowing anyway.

**Comment by bbbales2** [comment]
Eh, alright, we can go with this. I'll document this so folks know. Woulda had to anyway.

**Comment by bbbales2** [comment]
Would it make sense to throw in an |x| > 1e300 check as well? Just to make this fix super hard to trigger?

**Comment by bgoodri** [comment]
In the gradient_of_f function, we don't know if it is bounded or unbounded
integration. Maybe we should be passing the bounds in so that we know
whether to check something like |x| > 1e300 ?

**Comment by bgoodri** [comment]
So, this literally happened:

> C++ exception with description "Error in function boost::math::quadrature::sinh_sinh<double>::integrate: The function you are trying to integrate does not go to zero at -infinity, and instead evaluates to 0" thrown in the test body.

**Comment by bgoodri** [comment]
This is starting to look awesome. Two questions:

1. What is up with the default tolerance?
2. Do you know how to do a bivariate integral or is that impossible?

There is now a commented out test for the normal_sufficient_lpdf which tries to integrate over the mean and the sum of squares, but I haven't been able to get it to build. Most of the other one-dimensional distributions work.

**Comment by bbbales2** [comment]
> |x| > 1e300

Scratch this. Let's just go with if f == 0, df == nan -> set df = 0. More tape is just more tape in this situation.

> The function you are trying to integrate does not go to zero at -infinity, and instead evaluates to 0

Hahaha, took me a few readings to catch that. I like it.

> What is up with the default tolerance?

4th of July

> Do you know how to do a bivariate integral or is that impossible?

At least in the C++ version of this, sure:

```
TEST(StanMath_integrate_1d, TestMultiNormal) {
  using stan::math::exp;
  using stan::math::integrate_1d;
  using stan::math::var;

  var mu = 9.0 / 5;
  var sigma = 13.0 / 7;
  AVEC theta = {sigma};
  double b = std::numeric_limits<double>::infinity();
  double a = -b;
  auto pdf = [](auto x, auto xc, auto theta, auto x_r, auto x_i,
                std::ostream &msgs) {
    return exp(-0.5 * stan::math::square(x / theta[0]) - 0.5 * stan::math::square(x_r[0] / theta[0])) / (2 * 3.14159265358979323846 * theta\
[0] * theta[0]);
  };
  auto outer = [&](auto x, auto xc, auto theta, auto x_r, auto x_i,
                  std::ostream &msgs) {
    return integrate_1d(pdf, a, b, theta, {x}, {}, msgs, 1e-8);
  };
  var I = integrate_1d(outer, a, b, theta, {}, {}, msgs, 1e-8);
  EXPECT_FLOAT_EQ(1, I.val());

  AVEC x = createAVEC(mu, sigma);
  VEC g;
  I.grad(x, g);
  EXPECT_FLOAT_EQ(1, 1 + g[0]);
  EXPECT_FLOAT_EQ(1, 1 + g[1]);
}
```

Presumably the domain would have to be square/cubic. I don't even think this is terribly inefficient if it makes sense to just get the quadrature points as the combo of all the 1d quadrature points. There's probly a better way. Did you have a need for this?

What's with these: EXPECT_FLOAT_EQ(1, 1 + g[0]); ?

Why not EXPECT_NEAR(0.0, g[0], 1e-8)?

Nice set o' tests btw, tnx. Pretty cool to see it working.

**Comment by bgoodri** [comment]
There's a commented out test that the normal_sufficient PDF integrates to 1, but you have to integrate the mean over the whole real line and the sum of squares over the positive reals. I couldn't get that to build yet.

About the `EXPECT_FLOAT_EQ` thing: You can't test relative accuracy at zero (which would be the case for gradients of identities), but you can add 1.0 to both arguments.

**Comment by bbbales2** [comment]
> I couldn't get that to build yet.

```
return exp(stan::math::normal_sufficient_lpdf(x, s, 10,
                    theta[0], theta[1]));
       }, 0, std::numeric_limits<double>::infinity(),
          theta, {}, {}, std::cout, 1e-8);
```

```
       }, 0, std::numeric_limits<double>::infinity(),
```

```
          0
```

0.0 compiles at least. zero != 0 != 0.0 seems like this integrator is coming along nicely.

**Comment by bbbales2** [comment]
(I assume it's casting to a var and something is actually broken with two variable integration)

**Comment by bgoodri** [comment]
The infs and NaNs were coming from our own functions under prob/ . No one
would be integrating (a function of) a _lpdf because they are unbounded and
the gradients diverge as the random variable goes to infinity or even
approaches a finite boundary. But if you integrate a PDF, it must integrate
to 1 and the gradients must integrate to 0. The problem was when you call
exp() on a _lpdf in order to see if it integrates to 1, then the autodiff
does stuff like 0 * inf without knowing what limit that was converging to.
I think catching nan inside gradient_of_f and changing it to zero when the
function itself underflows is fine in this context.

**Comment by syclik** [comment]
Cool. That makes sense.

NaN's are coming out of our functions with good arguments? That sounds bad.

**Comment by bgoodri** [comment]
Extreme but good arguments. Like for the normal distribution, the
derivative with respect to sigma overflows before the function does.

**Comment by syclik** [comment]
Got it. Yes, at some point we tried checking on the propagation of the
derivatives, but I think we actually saw an effect on computation time when
it was done for every chain method. So... I think we're going to have to
live with that, unless there are better ways of checking.

**Comment by bob-carpenter** [comment]
The propagation of NaN between values and derivagives was in forward mode and it's a huge overhead that we need to get rid of if we ever get serious about forward mode.

**Comment by bbbales2** [comment]
@bgoodri I had a look at the tests you added. They seem good enough. I approve. Are you going to finish the commented out ones? If not, remove em' and let's get this thing merged.

**Pull request merged**